### ISSUE TITLE
Invalid cast from 'Double' to 'DateTime' after upgrade from DMS 0.9.1 to 0.9.3

### ISSUE THREAD

**Issue by VagueGit** [Mimetis/Dotmim.Sync#698]
Syncs most things then gets stuck at 
`91.67%: TableChangesApplyingSyncRowsArgs:Applying [SubscriptionTransactions] batch rows. State:Modified. Count:1`

The exception is `Invalid cast from 'Double' to 'DateTime'`.

The SQLite table looks like this
```
CREATE TABLE [SubscriptionTransactions] (
	"ItemValue"		numeric NOT NULL CHECK (ItemValue BETWEEN -9999998.99 AND 9999999.99),
	"OrderNr"		integer NOT NULL,
	"TransactionCredit"		numeric NOT NULL CHECK (TransactionCredit BETWEEN -9999998.99 AND 9999999.99),
	"TransactionDate"		date NOT NULL COLLATE NOCASE,
	"TransactionDebit"		numeric NOT NULL CHECK (TransactionDebit BETWEEN -9999998.99 AND 9999999.99),
	"TransactionDescription"		nvarchar(20) NOT NULL COLLATE NOCASE CHECK (length(TransactionDescription) <= 20),
	"TransactionId"		integer PRIMARY KEY AUTOINCREMENT NOT NULL,
	"TransactionType"		nvarchar(15) NOT NULL COLLATE NOCASE CHECK (length(TransactionType) <= 15)

);
```
The SQL Server table looks like this
```
CREATE TABLE [dbo].[SubscriptionTransactions](
	[ItemValue] [decimal](9, 2) NOT NULL,
	[OrderNr] [int] NOT NULL,
	[TransactionCredit] [decimal](9, 2) NOT NULL,
	[TransactionDate] [date] NOT NULL,
	[TransactionDebit] [decimal](9, 2) NOT NULL,
	[TransactionDescription] [nvarchar](20) NOT NULL,
	[TransactionId] [bigint] IDENTITY(1,1) NOT NULL,
	[TransactionType] [nvarchar](15) NOT NULL,
 CONSTRAINT [PK_SubscriptionTransactions] PRIMARY KEY CLUSTERED 
(
	[TransactionId] ASC
)
```
The batch looks like this
```
"changes": {
  "t": [
    {
      "n": "SubscriptionTransactions",
      "r": [
        [
          16,
          5,
          440.0,
          2,
          0.0,
          "2022-02-20T00:00:00",
          4.38,
          "2",
          "Job"
        ]
      ]
    }
  ]
},
```
The stack trace is
```
   at Dotmim.Sync.LocalOrchestrator.<ApplyChangesAsync>d__7.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Dotmim.Sync.SyncAgent.<>c__DisplayClass53_0.<<SynchronizeAsync>b__0>d.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Dotmim.Sync.SyncAgent.<SynchronizeAsync>d__53.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Wbl.Services.DataService.<SyncAsync>d__36.MoveNext() in C:\Repos\FREE\Wbl\Services\DataService.cs:line 342
```

On the client
```
DMS.SQLite 0.9.3
DMS.Web.Client 0.9.3
```
On the server
```
DMS.SqlServer.ChangeTracking 0.9.3
DMS.Web.Server 0.9.3
```

This is in our development environment

In production it's working fine using  DMS 0.9.1

**Comment by Mimetis** [comment]
batch is not looking good...
Is it a batch file from a snapshot generated from the version `0.9.1` ?
Or is it a batch generated on the fly (like a regular sync process) ?

**Comment by Mimetis** [comment]
I 've used this script:

``` sql
USE [master]
GO
-- Client database. No need to create the schema, Dotmim.Sync will do
if (exists (select * from sys.databases where name = 'vaguegit'))
Begin
	ALTER DATABASE vaguegit SET  SINGLE_USER WITH ROLLBACK IMMEDIATE;
	DROP DATABASE vaguegit
End
Create database vaguegit
GO

ALTER DATABASE vaguegit SET CHANGE_TRACKING = ON (CHANGE_RETENTION = 14 DAYS, AUTO_CLEANUP = ON)


Use vaguegit
GO

CREATE TABLE [dbo].[SubscriptionTransactions](
	[ItemValue] [decimal](9, 2) NOT NULL,
	[OrderNr] [int] NOT NULL,
	[TransactionCredit] [decimal](9, 2) NOT NULL,
	[TransactionDate] [date] NOT NULL,
	[TransactionDebit] [decimal](9, 2) NOT NULL,
	[TransactionDescription] [nvarchar](20) NOT NULL,
	[TransactionId] [bigint] IDENTITY(1,1) NOT NULL,
	[TransactionType] [nvarchar](15) NOT NULL,
 CONSTRAINT [PK_SubscriptionTransactions] PRIMARY KEY CLUSTERED 
(
	[TransactionId] ASC
))

Insert into SubscriptionTransactions
Values (1.0, 1, 100, getdate(), 55, 'desc', 'Debit')
```

And the code is pretty simple, just creating the sqlite on the fly

Everything is working fine
The batch generated looks like that:

``` json
  "changes": {
    "t": [
      {
        "n": "SubscriptionTransactions",
        "s": "",
        "r": [
          [
            16,
            1.0,
            1,
            100.0,
            "2022-02-22T00:00:00",
            55.0,
            "desc",
            1,
            "Debit"
          ]
        ]
      }
    ]
  },
```

If you look carefully, the order of the columns are not good in your case
That's what I'm trying to understand here.

**Comment by Mimetis** [comment]
Yes, i've got it.
You have your primary keys that are not ordered (I mean, usually primary keys are the first columns :) )
So far, the stored procedures are not generated in the correct order (order is important in v0.9.3)

In your table the column [TransactionId] is order number 7 (instead of 1)
And the stored procedure is generated like this (the _getchanges stored procedure):

``` sql
SELECT DISTINCT	[base].[ItemValue], 
	**[side].[TransactionId],**  <-----------------------------
	[base].[OrderNr], 
	[base].[TransactionCredit], 
	[base].[TransactionDate], 
	[base].[TransactionDebit], 
	[base].[TransactionDescription], 
	[base].[TransactionType], 
	[side].[sync_row_is_tombstone] as [sync_row_is_tombstone], 
	[side].[update_scope_id] as [sync_update_scope_id]
FROM [SubscriptionTransactions] [base]
RIGHT JOIN
```
And it should be generated like this:

``` sql
SELECT DISTINCT	[base].[ItemValue], 
	[base].[OrderNr], 
	[base].[TransactionCredit], 
	[base].[TransactionDate], 
	[base].[TransactionDebit], 
	[base].[TransactionDescription], 
	**[side].[TransactionId],** <-----------------------------
	[base].[TransactionType], 
	[side].[sync_row_is_tombstone] as [sync_row_is_tombstone], 
	[side].[update_scope_id] as [sync_update_scope_id]
FROM [SubscriptionTransactions] [base]
RIGHT JOIN
```

I'm currently doing the fix for the version `v0.9.4` 
I will commit the code in the master branch as soon as possible :)

**Comment by Mimetis** [comment]
It's resolved on the next version.
You have to Provision the stored procedures again to be sure you have the last correct version
